### PR TITLE
DNS overrides for 'awsvpc' tasks

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Apply DNS overrides in container's hostconfig

### Implementation details

If the container belongs to a task with an ENI, and if the ENI object has custom DNS overrides specified, override it in the container's host config as well.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: Yes

### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/api/task.go [hunk 1]
```diff
@@ -590,6 +596,27 @@ func (task *Task) shouldOverrideNetworkMode(container *Container, dockerContaine
...
+// 2. ENI has custom DNS IPs and search list associated with it
+// This should only be done for the pause container as other containers inherit
+// /etc/resolv.conf of this container (they share the network namespace)
+func (task *Task) overrideDNS(hostConfig *docker.HostConfig) *docker.HostConfig {
+	eni := task.GetTaskENI()
+	if eni == nil {
+		return hostConfig
+	}
+	if len(eni.DomainNameServers) > 0 {
+		hostConfig.DNS = eni.DomainNameServers
```
Comment (richardpen) [inline agent/api/task.go]: Not sure if we need add logs for these two check, if not, then do we still need the check?

File: agent/statemanager/state_manager.go [hunk 2]
```diff
@@ -50,7 +50,8 @@ const (
 	//   d) Add 'ENIAttachments' struct
 	//   e) Deprecate 'SteadyStateDependencies' in favor of 'TransitionDependencySet'
 	// 7)
-	//	 a) Add 'MetadataUpdated' field to 'api.Container'
+	//   a) Add 'MetadataUpdated' field to 'api.Container'
+	//   b) Add 'DomainNameServers' and 'DomainNameSearchList' in `api.ENI`
```
Comment (richardpen) [inline agent/statemanager/state_manager.go]: Are we sure that this is going together with metadata?

File: agent/api/task.go (hunk 1 above)
Comment (aaithal) [reply agent/api/task.go]: I don't think we need to log anything here. Why wouldn't we need these checks though?

File: agent/statemanager/state_manager.go (hunk 2 above)
Comment (aaithal) [reply agent/statemanager/state_manager.go]: We'll see :) If not, its a merge conflict that'll need resolution

File: agent/api/task.go (hunk 1 above)
Comment (richardpen) [reply agent/api/task.go]: Won't this has the same effect, we can even assign the empty value when the `len(xxx)==0`, any side effect?
```
hostConfig.DNS = eni.DomainNameServers
hostConfig.DNSSearch = eni.DomainNameSearchList
```

Comment (aaithal) [reply agent/api/task.go]: Oh I didn't consider that. Yeah, that's effectively doing the same thing. I'll update this.

**Comment by aaithal** [comment]
@vsiddharth oops! should be fixed now. Thanks for catching it.

**Pull request merged**